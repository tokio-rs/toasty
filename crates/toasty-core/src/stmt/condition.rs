use crate::stmt::{Expr, Node, Statement, Visit, VisitMut};

#[derive(Debug, Default, Clone, PartialEq)]
pub struct Condition {
    pub expr: Option<Expr>,
}

impl Condition {
    pub fn new(expr: impl Into<Expr>) -> Condition {
        Condition {
            expr: Some(expr.into()),
        }
    }

    pub fn is_some(&self) -> bool {
        self.expr.is_some()
    }

    pub fn is_none(&self) -> bool {
        self.expr.is_none()
    }
}

impl Statement {
    pub fn condition(&self) -> Option<&Condition> {
        match self {
            Statement::Update(update) if update.condition.is_some() => Some(&update.condition),
            _ => None,
        }
    }

    /// Returns a mutable reference to the statement's condition.
    ///
    /// Returns `None` for statements that do not support conditions.
    pub fn condition_mut(&mut self) -> Option<&mut Condition> {
        match self {
            Statement::Update(update) => Some(&mut update.condition),
            _ => None,
        }
    }

    /// Returns a mutable reference to the statement's condition.
    ///
    /// # Panics
    ///
    /// Panics if the statement does not support conditions.
    #[track_caller]
    pub fn condition_mut_unwrap(&mut self) -> &mut Condition {
        match self {
            Statement::Update(update) => &mut update.condition,
            _ => panic!("expected Statement with condition"),
        }
    }
}

impl Node for Condition {
    fn visit<V: Visit>(&self, mut visit: V) {
        visit.visit_condition(self);
    }

    fn visit_mut<V: VisitMut>(&mut self, mut visit: V) {
        visit.visit_condition_mut(self);
    }
}

impl<T> From<T> for Condition
where
    Expr: From<T>,
{
    fn from(value: T) -> Self {
        Condition {
            expr: Some(value.into()),
        }
    }
}
