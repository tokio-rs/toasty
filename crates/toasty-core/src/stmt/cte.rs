use super::Query;

#[derive(Debug, Clone)]
pub struct Cte {
    pub query: Query,
}
