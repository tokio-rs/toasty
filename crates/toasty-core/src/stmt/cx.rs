use crate::{
    schema::app::{Field, Model, ModelId},
    stmt::{ExprReference, ExprSet, InsertTarget, Query, Select, Source, Update, UpdateTarget},
    Schema,
};

pub struct ExprContext<'a> {
    schema: &'a Schema,
    parent: Option<&'a ExprContext<'a>>,
    target: ExprTarget<'a>,
}

#[derive(Debug, Clone, Copy)]
pub enum ExprTarget<'a> {
    Const,
    Insert(&'a InsertTarget),
    Source(&'a Source),
    Update(&'a UpdateTarget),
}

impl<'a> ExprContext<'a> {
    pub fn new(schema: &'a Schema) -> ExprContext<'a> {
        ExprContext::new_with_target(schema, ExprTarget::Const)
    }

    pub fn new_with_target(
        schema: &'a Schema,
        target: impl Into<ExprTarget<'a>>,
    ) -> ExprContext<'a> {
        ExprContext {
            schema,
            parent: None,
            target: target.into(),
        }
    }

    pub fn schema(&self) -> &'a Schema {
        self.schema
    }

    pub fn target(&self) -> ExprTarget<'a> {
        self.target
    }

    pub fn scope<'child>(
        &'child self,
        target: impl Into<ExprTarget<'child>>,
    ) -> ExprContext<'child> {
        ExprContext {
            schema: self.schema,
            parent: Some(self),
            target: target.into(),
        }
    }

    pub fn target_as_model(&self) -> Option<&'a Model> {
        let model_id = self.target.as_model_id()?;
        Some(self.schema.app.model(model_id))
    }

    pub fn resolve_expr_reference(&self, expr_reference: &ExprReference) -> &'a Field {
        let ExprReference::Field { nesting, index } = expr_reference else {
            todo!();
        };

        let mut curr = self;

        // Walk up the stack
        for _ in 0..*nesting {
            let Some(parent) = self.parent else {
                todo!("bug");
            };

            curr = parent;
        }

        match curr.target {
            ExprTarget::Const => todo!("fail"),
            ExprTarget::Source(Source::Model(source_model)) => {
                &self.schema.app.model(source_model.model).fields[*index]
            }
            ExprTarget::Source(_) => todo!(),
            ExprTarget::Insert(InsertTarget::Model(model_id)) => {
                &self.schema.app.model(model_id).fields[*index]
            }
            ExprTarget::Insert(_) => todo!(),
            ExprTarget::Update(UpdateTarget::Query(query)) => {
                let model_id = query.body.as_select().source.as_model_id();
                &self.schema.app.model(model_id).fields[*index]
            }
            ExprTarget::Update(UpdateTarget::Model(model_id)) => {
                &self.schema.app.model(model_id).fields[*index]
            }
            ExprTarget::Update(UpdateTarget::Table(_)) => todo!(),
        }
    }
}

impl<'a> ExprTarget<'a> {
    pub fn as_model_id(self) -> Option<ModelId> {
        Some(match self {
            ExprTarget::Source(Source::Model(source_model)) => source_model.model,
            ExprTarget::Update(UpdateTarget::Model(model_id)) => *model_id,
            ExprTarget::Insert(InsertTarget::Model(model_id)) => *model_id,
            _ => return None,
        })
    }
}

impl<'a> From<&'a Query> for ExprTarget<'a> {
    fn from(value: &'a Query) -> Self {
        ExprTarget::from(&value.body)
    }
}

impl<'a> From<&'a ExprSet> for ExprTarget<'a> {
    fn from(value: &'a ExprSet) -> Self {
        match value {
            ExprSet::Select(select) => ExprTarget::from(&**select),
            ExprSet::SetOp(_) => todo!(),
            ExprSet::Update(update) => ExprTarget::from(&**update),
            ExprSet::Values(_) => ExprTarget::Const,
        }
    }
}

impl<'a> From<&'a Select> for ExprTarget<'a> {
    fn from(value: &'a Select) -> Self {
        ExprTarget::from(&value.source)
    }
}

impl<'a> From<&'a InsertTarget> for ExprTarget<'a> {
    fn from(value: &'a InsertTarget) -> Self {
        ExprTarget::Insert(value)
    }
}

impl<'a> From<&'a Source> for ExprTarget<'a> {
    fn from(value: &'a Source) -> Self {
        ExprTarget::Source(value)
    }
}

impl<'a> From<&'a Update> for ExprTarget<'a> {
    fn from(value: &'a Update) -> Self {
        ExprTarget::from(&value.target)
    }
}

impl<'a> From<&'a UpdateTarget> for ExprTarget<'a> {
    fn from(value: &'a UpdateTarget) -> Self {
        ExprTarget::Update(value)
    }
}
