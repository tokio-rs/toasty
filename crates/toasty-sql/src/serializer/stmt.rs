use std::mem;

use super::{Comma, Delimited, Ident, Params, ToSql};

use crate::{serializer::ExprContext, stmt};
use toasty_core::{schema::db, stmt::SourceTableId};

struct ColumnsWithConstraints<'a>(&'a stmt::CreateTable);

impl ToSql for ColumnsWithConstraints<'_> {
    fn to_sql<P: Params>(self, cx: &ExprContext<'_>, f: &mut super::Formatter<'_, P>) {
        let columns = Comma(&self.0.columns);

        if let Some(pk) = &self.0.primary_key {
            fmt!(cx, f, columns ", PRIMARY KEY " pk);
        } else {
            fmt!(cx, f, columns);
        }
    }
}

impl ToSql for &stmt::CreateIndex {
    fn to_sql<P: Params>(self, cx: &ExprContext<'_>, f: &mut super::Formatter<'_, P>) {
        let index = f.serializer.index(self.index);
        let table = f.serializer.table(self.on);
        let index_name = Ident(&index.name);
        let table_name = Ident(&table.name);
        let columns = Comma(&self.columns);
        let unique = if self.unique { "UNIQUE " } else { "" };

        // Create a new expression scope to serialize the statement
        let cx = cx.scope(table);

        fmt!(
            &cx, f, "CREATE " unique "INDEX " index_name " ON " table_name " (" columns ")"
        );
    }
}

impl ToSql for &stmt::CreateTable {
    fn to_sql<P: Params>(self, cx: &ExprContext<'_>, f: &mut super::Formatter<'_, P>) {
        let table = f.serializer.table(self.table);
        let name = Ident(&table.name);
        let columns = ColumnsWithConstraints(self);

        // Create new expression scope to serialize the statement
        let cx = cx.scope(table);

        fmt!(
            &cx, f, "CREATE TABLE " name " (" columns ")"
        );
    }
}

impl ToSql for &stmt::Delete {
    fn to_sql<P: Params>(self, cx: &ExprContext<'_>, f: &mut super::Formatter<'_, P>) {
        let prev = mem::replace(&mut f.alias, true);

        assert!(self.returning.is_none());

        // Create a new expression scope to serialize the statement
        let cx = cx.scope(self);

        fmt!(&cx, f, "DELETE FROM " self.from " WHERE " self.filter);

        f.alias = prev;
    }
}

impl ToSql for &stmt::Direction {
    fn to_sql<P: Params>(self, cx: &ExprContext<'_>, f: &mut super::Formatter<'_, P>) {
        match self {
            stmt::Direction::Asc => fmt!(cx, f, "ASC"),
            stmt::Direction::Desc => fmt!(cx, f, "DESC"),
        }
    }
}

impl ToSql for &stmt::DropTable {
    fn to_sql<P: Params>(self, cx: &ExprContext<'_>, f: &mut super::Formatter<'_, P>) {
        let if_exists = if self.if_exists { "IF EXISTS " } else { "" };
        fmt!(cx, f, "DROP TABLE " if_exists self.name);
    }
}

impl ToSql for &stmt::Insert {
    fn to_sql<P: Params>(self, cx: &ExprContext<'_>, f: &mut super::Formatter<'_, P>) {
        // Create a new expression scope to serialize the statement
        let cx = cx.scope(self);

        let returning = self
            .returning
            .as_ref()
            .map(|returning| ("RETURNING ", returning));

        fmt!(
            &cx, f, "INSERT INTO " self.target " " self.source returning
        );
    }
}

impl ToSql for &stmt::InsertTarget {
    fn to_sql<P: Params>(self, cx: &ExprContext<'_>, f: &mut super::Formatter<'_, P>) {
        match self {
            stmt::InsertTarget::Table(insert_table) => {
                let table_name = f.serializer.table_name(insert_table);
                let columns = Comma(
                    insert_table
                        .columns
                        .iter()
                        .map(|column_id| f.serializer.column_name(*column_id)),
                );

                fmt!(cx, f, table_name " (" columns ")");
            }
            _ => todo!("self={self:?}"),
        }
    }
}

impl ToSql for &stmt::Limit {
    fn to_sql<P: Params>(self, cx: &ExprContext<'_>, f: &mut super::Formatter<'_, P>) {
        assert!(self.offset.is_none(), "TODO");

        fmt!(cx, f, "LIMIT " self.limit);
    }
}

impl ToSql for &stmt::Query {
    fn to_sql<P: Params>(self, cx: &ExprContext<'_>, f: &mut super::Formatter<'_, P>) {
        let prev = mem::replace(&mut f.alias, true);
        // Create a new expression scope to serialize the statement
        let cx = cx.scope(self);

        let locks = if self.locks.is_empty() {
            None
        } else {
            Some((" ", Delimited(&self.locks, " ")))
        };

        let body = &self.body;
        let order_by = self.order_by.as_ref().map(|order_by| (" ", order_by));
        let limit = self.limit.as_ref().map(|limit| (" ", limit));

        fmt!(&cx, f, self.with body order_by limit locks);

        f.alias = prev;
    }
}

impl ToSql for &stmt::ExprSet {
    fn to_sql<P: Params>(self, cx: &ExprContext<'_>, f: &mut super::Formatter<'_, P>) {
        match self {
            stmt::ExprSet::Select(expr) => expr.to_sql(cx, f),
            stmt::ExprSet::Values(expr) => expr.to_sql(cx, f),
            stmt::ExprSet::Update(expr) => expr.to_sql(cx, f),
            _ => todo!("self={self:?}"),
        }
    }
}

impl ToSql for &stmt::OrderBy {
    fn to_sql<P: Params>(self, cx: &ExprContext<'_>, f: &mut super::Formatter<'_, P>) {
        let order_by = Comma(&self.exprs);

        fmt!(cx, f, "ORDER BY " order_by);
    }
}

impl ToSql for &stmt::OrderByExpr {
    fn to_sql<P: Params>(self, cx: &ExprContext<'_>, f: &mut super::Formatter<'_, P>) {
        if let Some(order) = &self.order {
            fmt!(cx, f, self.expr " " order);
        } else {
            fmt!(cx, f, self.expr);
        }
    }
}

impl ToSql for &stmt::Returning {
    fn to_sql<P: Params>(self, cx: &ExprContext<'_>, f: &mut super::Formatter<'_, P>) {
        match self {
            stmt::Returning::Model { .. } => fmt!(cx, f, "*"),
            stmt::Returning::Expr(stmt::Expr::Record(expr_record)) => {
                let fields = expr_record
                    .fields
                    .iter()
                    .enumerate()
                    .map(|(i, expr)| match expr {
                        stmt::Expr::Column(_) => (expr, None, None),
                        _ => (expr, Some(" AS col_"), Some(i)),
                    });

                fmt!(cx, f, Comma(fields));
            }
            stmt::Returning::Expr(expr) => {
                fmt!(cx, f, expr);
            }
            _ => todo!(),
        }
    }
}

impl ToSql for &stmt::Select {
    fn to_sql<P: Params>(self, cx: &ExprContext<'_>, f: &mut super::Formatter<'_, P>) {
        fmt!(
            cx, f,
            "SELECT " self.returning " FROM " self.source
            " WHERE " self.filter
        );
    }
}

impl ToSql for &stmt::Lock {
    fn to_sql<P: Params>(self, cx: &ExprContext<'_>, f: &mut super::Formatter<'_, P>) {
        match self {
            stmt::Lock::Update => fmt!(cx, f, "FOR UPDATE"),
            stmt::Lock::Share => fmt!(cx, f, "FOR SHARE"),
        }
    }
}

impl ToSql for &stmt::Source {
    fn to_sql<P: Params>(self, cx: &ExprContext<'_>, f: &mut super::Formatter<'_, P>) {
        match self {
            stmt::Source::Table(source_table) => {
                source_table.to_sql(cx, f);
            }
            _ => todo!("self={self:?}"),
        }
    }
}

impl ToSql for &toasty_core::stmt::Statement {
    fn to_sql<P: Params>(self, cx: &ExprContext<'_>, f: &mut super::Formatter<'_, P>) {
        use toasty_core::stmt::Statement::*;

        f.depth += 1;

        match self {
            Delete(stmt) => stmt.to_sql(cx, f),
            Insert(stmt) => stmt.to_sql(cx, f),
            Query(stmt) => stmt.to_sql(cx, f),
            Update(stmt) => stmt.to_sql(cx, f),
        }

        f.depth -= 1;
    }
}

impl ToSql for &stmt::Statement {
    fn to_sql<P: Params>(self, cx: &ExprContext<'_>, f: &mut super::Formatter<'_, P>) {
        match self {
            stmt::Statement::CreateIndex(stmt) => stmt.to_sql(cx, f),
            stmt::Statement::CreateTable(stmt) => stmt.to_sql(cx, f),
            stmt::Statement::DropTable(stmt) => stmt.to_sql(cx, f),
            stmt::Statement::Delete(stmt) => stmt.to_sql(cx, f),
            stmt::Statement::Insert(stmt) => stmt.to_sql(cx, f),
            stmt::Statement::Query(stmt) => stmt.to_sql(cx, f),
            stmt::Statement::Update(stmt) => stmt.to_sql(cx, f),
        }
    }
}

impl ToSql for &stmt::SourceTable {
    fn to_sql<P: Params>(self, cx: &ExprContext<'_>, f: &mut super::Formatter<'_, P>) {
        // Serialize the main table relation
        match &self.from_item.relation {
            stmt::TableFactor::Table(table_id) => {
                let table_ref = &self.tables[table_id.0];
                let alias = TableAlias {
                    depth: f.depth,
                    table: *table_id,
                };

                fmt!(cx, f, table_ref " AS " alias);
            }
        }

        // Serialize the joins
        for join in &self.from_item.joins {
            match &join.constraint {
                stmt::JoinOp::Left(expr) => {
                    let join_table_ref = &self.tables[join.table.0];
                    let alias = TableAlias {
                        depth: f.depth,
                        table: join.table,
                    };
                    fmt!(cx, f, " LEFT JOIN " join_table_ref " AS " alias " ON " expr);
                }
            }
        }
    }
}

impl ToSql for &stmt::TableRef {
    fn to_sql<P: Params>(self, cx: &ExprContext<'_>, f: &mut super::Formatter<'_, P>) {
        match *self {
            stmt::TableRef::Table(table_id) => {
                let table_name = f.serializer.table_name(table_id);
                fmt!(cx, f, table_name);
            }
            stmt::TableRef::Cte { nesting, index } => {
                assert!(f.depth >= nesting, "nesting={nesting} depth={}", f.depth);

                let depth = f.depth - nesting;
                fmt!(cx, f, "cte_" depth "_" index);
            }
        }
    }
}

struct TableAlias {
    depth: usize,
    table: SourceTableId,
}

impl ToSql for &TableAlias {
    fn to_sql<P: Params>(self, cx: &ExprContext<'_>, f: &mut super::Formatter<'_, P>) {
        fmt!(cx, f, "tbl_" self.depth "_" self.table.0);
    }
}

impl ToSql for &stmt::Update {
    fn to_sql<P: Params>(self, cx: &ExprContext<'_>, f: &mut super::Formatter<'_, P>) {
        let prev = mem::replace(&mut f.alias, true);

        let table = f.serializer.schema.table(self.target.as_table());
        let assignments = (table, &self.assignments);

        // Create a new expression scope to serialize the statement
        let cx = cx.scope(self);

        let filter = self.filter.as_ref().map(|expr| (" WHERE ", expr));
        let returning = self
            .returning
            .as_ref()
            .map(|returning| (" RETURNING ", returning));

        assert!(
            self.condition.is_none(),
            "SQL does not support update conditions"
        );

        fmt!(&cx, f, "UPDATE " self.target " SET " assignments filter returning);

        f.alias = prev;
    }
}

impl ToSql for (&db::Table, &stmt::Assignments) {
    fn to_sql<P: Params>(self, cx: &ExprContext<'_>, f: &mut super::Formatter<'_, P>) {
        let frags = self.1.iter().map(|(index, assignment)| {
            let column_name = Ident(&self.0.columns[index].name);
            (column_name, " = ", &assignment.expr)
        });

        fmt!(cx, f, Delimited(frags, ", "));
    }
}

impl ToSql for &stmt::UpdateTarget {
    fn to_sql<P: Params>(self, cx: &ExprContext<'_>, f: &mut super::Formatter<'_, P>) {
        match self {
            stmt::UpdateTarget::Table(table_id) => {
                let table_name = f.serializer.table_name(*table_id);
                let alias = TableAlias {
                    depth: f.depth,
                    table: SourceTableId(0),
                };

                fmt!(cx, f, table_name " AS " alias);
            }
            _ => todo!(),
        }
    }
}

impl ToSql for &stmt::Values {
    fn to_sql<P: Params>(self, cx: &ExprContext<'_>, f: &mut super::Formatter<'_, P>) {
        let rows = Comma(self.rows.iter());

        fmt!(cx, f, "VALUES " rows)
    }
}
