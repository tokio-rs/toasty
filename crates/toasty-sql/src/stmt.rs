mod column_def;
pub use column_def::ColumnDef;

mod create_index;
pub use create_index::CreateIndex;

mod create_table;
pub use create_table::CreateTable;

mod drop_table;
pub use drop_table::DropTable;

mod name;
pub use name::Name;

pub use toasty_core::stmt::*;

#[derive(Debug, Clone)]
pub enum Statement {
    CreateIndex(CreateIndex),
    CreateTable(CreateTable),
    DropTable(DropTable),
    Delete(Delete),
    Insert(Insert),
    Query(Query),
    Update(Update),
}

impl Statement {
    pub fn is_update(&self) -> bool {
        matches!(self, Self::Update(_))
    }

    /// Returns the number of returned elements within the statement (if one exists).
    pub fn returning_len(&self) -> Option<usize> {
        match self {
            Self::Delete(delete) => delete
                .returning
                .as_ref()
                .map(|ret| ret.as_expr().as_record().len()),
            Self::Insert(insert) => insert
                .returning
                .as_ref()
                .map(|ret| ret.as_expr().as_record().len()),
            Self::Query(query) => match &*query.body {
                ExprSet::Select(select) => Some(select.returning.as_expr().as_record().len()),
                stmt => todo!("returning_len, stmt={stmt:#?}"),
            },
            Self::Update(update) => update
                .returning
                .as_ref()
                .map(|ret| ret.as_expr().as_record().len()),
            _ => None,
        }
    }
}

impl From<toasty_core::stmt::Statement> for Statement {
    fn from(value: toasty_core::stmt::Statement) -> Self {
        match value {
            toasty_core::stmt::Statement::Delete(stmt) => Self::Delete(stmt),
            toasty_core::stmt::Statement::Insert(stmt) => Self::Insert(stmt),
            toasty_core::stmt::Statement::Query(stmt) => Self::Query(stmt),
            toasty_core::stmt::Statement::Update(stmt) => Self::Update(stmt),
        }
    }
}
