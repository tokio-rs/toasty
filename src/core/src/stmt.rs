mod assignments;
pub use assignments::{Assignment, AssignmentOp, Assignments};

mod association;
pub use association::Association;

mod delete;
pub use delete::Delete;

mod direction;
pub use direction::Direction;

mod entry;
pub use entry::Entry;

mod entry_mut;
pub use entry_mut::EntryMut;

mod entry_path;
pub use entry_path::EntryPath;

mod expr;
pub use expr::Expr;

mod expr_and;
pub use expr_and::ExprAnd;

mod expr_arg;
pub use expr_arg::ExprArg;

mod expr_begins_with;
pub use expr_begins_with::ExprBeginsWith;

mod expr_binary_op;
pub use expr_binary_op::ExprBinaryOp;

mod expr_cast;
pub use expr_cast::ExprCast;

mod expr_column;
pub use expr_column::ExprColumn;

mod expr_concat;
pub use expr_concat::ExprConcat;

mod expr_concat_str;
pub use expr_concat_str::ExprConcatStr;

mod expr_enum;
pub use expr_enum::ExprEnum;

mod expr_field;
pub use expr_field::ExprField;

mod expr_in_list;
pub use expr_in_list::ExprInList;

mod expr_in_subquery;
pub use expr_in_subquery::ExprInSubquery;

mod expr_is_null;
pub use expr_is_null::ExprIsNull;

mod expr_key;
pub use expr_key::ExprKey;

mod expr_like;
pub use expr_like::ExprLike;

mod expr_list;
pub use expr_list::ExprList;

mod expr_map;
pub use expr_map::ExprMap;

mod expr_or;
pub use expr_or::ExprOr;

mod expr_order_by;
pub use expr_order_by::ExprOrderBy;

mod expr_pattern;
pub use expr_pattern::ExprPattern;

mod expr_project;
pub use expr_project::ExprProject;

mod expr_record;
pub use expr_record::ExprRecord;

mod expr_set;
pub use expr_set::ExprSet;

mod expr_set_op;
pub use expr_set_op::ExprSetOp;

mod expr_stmt;
pub use expr_stmt::ExprStmt;

mod expr_ty;
pub use expr_ty::ExprTy;

mod id;
pub use id::Id;

mod insert;
pub use insert::Insert;

mod insert_table;
pub use insert_table::InsertTable;

mod insert_target;
pub use insert_target::InsertTarget;

mod node;
pub use node::Node;

mod op_binary;
pub use op_binary::BinaryOp;

mod op_set;
pub use op_set::SetOp;

mod path;
pub use path::Path;

mod path_field_set;
pub use path_field_set::PathFieldSet;

mod projection;
pub use projection::Projection;

mod query;
pub use query::Query;

mod returning;
pub use returning::Returning;

mod select;
pub use select::Select;

mod source;
pub use source::{Source, SourceModel};

mod sparse_record;
pub use sparse_record::SparseRecord;

pub mod sql;

pub mod substitute;

mod table_with_joins;
pub use table_with_joins::TableWithJoins;

mod ty;
pub use ty::Type;

mod ty_enum;
pub use ty_enum::{EnumVariant, TypeEnum};

mod update;
pub use update::{Update, UpdateTarget};

mod value;
pub use value::Value;

mod values;
pub use values::Values;

mod value_enum;
pub use value_enum::ValueEnum;

mod value_record;
pub use value_record::ValueRecord;

pub mod visit_mut;
pub use visit_mut::VisitMut;

mod value_stream;
pub use value_stream::ValueStream;

pub mod visit;
pub use visit::Visit;

use crate::{
    schema::{
        app::{Field, FieldId, Model, ModelId},
        db::{Column, ColumnId, TableId},
    },
    stmt, Error, Result,
};

use std::fmt;

#[derive(Clone, PartialEq)]
pub enum Statement {
    /// Delete one or more existing records
    Delete(Delete),

    /// Create one or more instances of a model
    Insert(Insert),

    /// Query the database
    Query(Query),

    /// Update one or more existing records
    Update(Update),
}

impl Statement {
    pub fn substitute(&mut self, mut input: impl substitute::Input) {
        match self {
            Statement::Query(stmt) => stmt.substitute_ref(&mut input),
            _ => todo!("stmt={self:#?}"),
        }
    }
}

impl Node for Statement {
    fn visit<V: Visit>(&self, mut visit: V) {
        visit.visit_stmt(self);
    }

    fn visit_mut<V: VisitMut>(&mut self, mut visit: V) {
        visit.visit_stmt_mut(self);
    }
}

impl fmt::Debug for Statement {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Statement::Delete(v) => v.fmt(f),
            Statement::Insert(v) => v.fmt(f),
            Statement::Query(v) => v.fmt(f),
            Statement::Update(v) => v.fmt(f),
        }
    }
}
