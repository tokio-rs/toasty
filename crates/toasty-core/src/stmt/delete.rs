use super::{Node, Query, Returning, Source, Statement, Visit, VisitMut};
use crate::stmt::{self, Filter};

#[derive(Debug, Clone, PartialEq)]
pub struct Delete {
    /// Source of data to delete from
    pub from: Source,

    /// WHERE
    pub filter: Filter,

    /// Optionally, return something
    pub returning: Option<Returning>,
}

impl Delete {
    pub fn selection(&self) -> Query {
        stmt::Query::new_select(self.from.model_id_unwrap(), self.filter.clone())
    }
}

impl From<Delete> for Statement {
    fn from(src: Delete) -> Self {
        Self::Delete(src)
    }
}

impl Node for Delete {
    fn visit<V: Visit>(&self, mut visit: V) {
        visit.visit_stmt_delete(self);
    }

    fn visit_mut<V: VisitMut>(&mut self, mut visit: V) {
        visit.visit_stmt_delete_mut(self);
    }
}
