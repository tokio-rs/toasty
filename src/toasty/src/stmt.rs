mod association;
pub use association::Association;

mod delete;
pub use delete::Delete;

mod expr;
pub use expr::Expr;

mod id;
pub use id::Id;

mod insert;
pub use insert::Insert;

mod into_expr;
pub use into_expr::IntoExpr;

mod into_insert;
pub use into_insert::IntoInsert;

mod into_select;
pub use into_select::IntoSelect;

mod path;
pub use path::Path;

mod select;
pub use select::Select;

mod to_statement;
pub use to_statement::ToStatement;

mod update;
pub use update::Update;

pub use toasty_core::stmt::Value;

use crate::Model;

use toasty_core::stmt;

use std::{fmt, marker::PhantomData};

pub struct Statement<M> {
    pub(crate) untyped: stmt::Statement,
    _p: PhantomData<M>,
}

impl<M: Model> Statement<M> {
    pub fn from_untyped(query: impl IntoSelect<Model = M>) -> Statement<M> {
        Statement {
            untyped: query.into_select().untyped.into(),
            _p: PhantomData,
        }
    }
}

impl<M> From<Select<M>> for Statement<M> {
    fn from(value: Select<M>) -> Self {
        Statement {
            untyped: value.untyped.into(),
            _p: PhantomData,
        }
    }
}

impl<M> From<Insert<M>> for Statement<M> {
    fn from(value: Insert<M>) -> Self {
        Statement {
            untyped: value.untyped.into(),
            _p: PhantomData,
        }
    }
}

impl<M> From<Update<M>> for Statement<M> {
    fn from(value: Update<M>) -> Self {
        Statement {
            untyped: value.untyped.into(),
            _p: PhantomData,
        }
    }
}

impl<M> fmt::Debug for Statement<M> {
    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.untyped.fmt(fmt)
    }
}
