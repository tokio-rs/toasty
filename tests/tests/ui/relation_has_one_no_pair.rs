use toasty::stmt::Id;

#[toasty::model]
struct User {
    #[key]
    #[auto]
    id: Id<Self>,

    #[has_one]
    profile: Option<Profile>,
}

#[toasty::model]
struct Profile {
    #[key]
    #[auto]
    id: Id<Self>,
}

fn main() {}