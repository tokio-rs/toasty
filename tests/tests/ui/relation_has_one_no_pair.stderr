error[E0609]: no field `user` on type `Profile`
 --> tests/ui/relation_has_one_no_pair.rs:9:5
  |
9 |     #[has_one]
  |     ^ unknown field
  |
  = note: available field is: `id`
